File: src/path_info.rs
use crate::platform::{self, PathStyle};

/// Priority of a path segment for truncation decisions.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum SegmentPriority {
    /// Filename: never shrink.
    Sacred = 0,
    /// User identity segment (username/profile name): preserve when possible.
    Identity = 1,
    /// Well-known application/service context: nice to have.
    Context = 2,
    /// Generic intermediate directory: first to be shortened.
    Expendable = 3,
}

/// A single path segment with its text and computed priority.
#[derive(Debug, Clone)]
pub struct Segment {
    pub text: String,
    pub priority: SegmentPriority,
}

/// Parsed representation of a path.
#[derive(Debug, Clone)]
pub struct PathInfo {
    /// Prefix: root (`/`, `C:\`, `\\server\share\`, `.`, `~`, or empty).
    pub prefix: String,
    /// Directory segments between prefix and filename.
    pub segments: Vec<Segment>,
    /// The final component (filename). Sacred, never truncated.
    pub filename: String,
    /// Detected or forced path style.
    pub style: PathStyle,
}

impl PathInfo {
    /// Parse a path string into structured components with priority classification.
    pub fn parse(path: &str, force_style: Option<PathStyle>) -> Self {
        let style = force_style.unwrap_or_else(|| platform::detect_style(path));
        let sep = platform::separator(style);

        // Normalize: replace the "other" separator with the canonical one
        let normalized: String = match style {
            PathStyle::Unix => path.replace('\\', "/"),
            PathStyle::Windows => path.replace('/', "\\"),
        };

        let (prefix, remainder) = extract_prefix(&normalized, style);

        // Split remaining path into parts
        let parts: Vec<&str> = remainder.split(sep).filter(|s| !s.is_empty()).collect();

        if parts.is_empty() {
            return PathInfo {
                prefix,
                segments: Vec::new(),
                filename: String::new(),
                style,
            };
        }

        // Last part is the filename (sacred)
        let filename = parts.last().unwrap().to_string();
        let dir_parts = &parts[..parts.len() - 1];

        // Classify segments
        let segments = classify_segments(dir_parts, &prefix, style);

        PathInfo {
            prefix,
            segments,
            filename,
            style,
        }
    }

    /// Reassemble the path from prefix + segments + filename using the given separator.
    pub fn reassemble(&self, segment_texts: &[&str]) -> String {
        let sep = platform::separator(self.style);
        let mut result = self.prefix.clone();

        for (i, text) in segment_texts.iter().enumerate() {
            if i > 0 || (!result.is_empty() && !result.ends_with(sep)) {
                result.push(sep);
            }
            result.push_str(text);
        }

        if !self.filename.is_empty() {
            if !result.is_empty() && !result.ends_with(sep) {
                result.push(sep);
            }
            result.push_str(&self.filename);
        }

        result
    }
}

/// Extract the prefix (root portion) from a normalized path.
fn extract_prefix(path: &str, style: PathStyle) -> (String, &str) {
    match style {
        PathStyle::Windows => {
            let bytes = path.as_bytes();
            // UNC: \\server\share
            if bytes.starts_with(b"\\\\") {
                let after_slashes = &path[2..];
                // Find server
                if let Some(server_end) = after_slashes.find('\\') {
                    let after_server = &after_slashes[server_end + 1..];
                    // Find share
                    let share_end = after_server.find('\\').unwrap_or(after_server.len());
                    let prefix_end = 2 + server_end + 1 + share_end;
                    let prefix = &path[..prefix_end];
                    let remainder = if prefix_end < path.len() {
                        &path[prefix_end + 1..]
                    } else {
                        ""
                    };
                    return (prefix.to_string(), remainder);
                }
                return (path.to_string(), "");
            }
            // Drive letter: C:\
            if bytes.len() >= 3
                && bytes[0].is_ascii_alphabetic()
                && bytes[1] == b':'
                && bytes[2] == b'\\'
            {
                return (path[..3].to_string(), &path[3..]);
            }
            // Dot-backslash: .\
            if bytes.starts_with(b".\\") {
                return (".".to_string(), &path[2..]);
            }
            // Just backslash
            if bytes.starts_with(b"\\") {
                return ("\\".to_string(), &path[1..]);
            }
            (String::new(), path)
        }
        PathStyle::Unix => {
            let bytes = path.as_bytes();
            // Absolute: /
            if bytes.starts_with(b"/") {
                return ("/".to_string(), &path[1..]);
            }
            // Tilde: ~/
            if bytes.starts_with(b"~/") {
                return ("~".to_string(), &path[2..]);
            }
            if path == "~" {
                return ("~".to_string(), "");
            }
            (String::new(), path)
        }
    }
}

/// Classify directory segments by priority.
fn classify_segments(parts: &[&str], prefix: &str, style: PathStyle) -> Vec<Segment> {
    let home_roots = match style {
        PathStyle::Unix => platform::UNIX_HOME_ROOTS,
        PathStyle::Windows => platform::WIN_HOME_ROOTS,
    };

    // Find the identity segment index: the segment AFTER a home root
    let identity_idx = parts.iter().enumerate().find_map(|(i, &part)| {
        if i == 0
            && home_roots
                .iter()
                .any(|&root| part.eq_ignore_ascii_case(root))
        {
            // The next segment (i+1) is the username
            if i + 1 < parts.len() {
                return Some(i + 1);
            }
        }
        None
    });

    // Special case: tilde prefix means no explicit identity segment in parts
    // (the ~ already encodes the user)

    parts
        .iter()
        .enumerate()
        .map(|(i, &text)| {
            let priority = if Some(i) == identity_idx {
                SegmentPriority::Identity
            } else if i == 0
                && home_roots
                    .iter()
                    .any(|&root| text.eq_ignore_ascii_case(root))
            {
                // "Users" or "home" segment itself — keep it (Context)
                SegmentPriority::Context
            } else if is_well_known_prefix(prefix) && i == 0 {
                // First segment after a well-known prefix like C:\Windows
                SegmentPriority::Context
            } else {
                SegmentPriority::Expendable
            };

            Segment {
                text: text.to_string(),
                priority,
            }
        })
        .collect()
}

/// Check if the prefix indicates a well-known system path.
fn is_well_known_prefix(_prefix: &str) -> bool {
    false // Conservative: don't over-classify
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_unix_home() {
        let info = PathInfo::parse("/home/john/projects/rust/file.rs", None);
        assert_eq!(info.prefix, "/");
        assert_eq!(info.filename, "file.rs");
        assert_eq!(info.segments.len(), 4);
        assert_eq!(info.segments[0].text, "home");
        assert_eq!(info.segments[0].priority, SegmentPriority::Context);
        assert_eq!(info.segments[1].text, "john");
        assert_eq!(info.segments[1].priority, SegmentPriority::Identity);
        assert_eq!(info.segments[2].priority, SegmentPriority::Expendable);
    }

    #[test]
    fn parse_windows_drive() {
        let info = PathInfo::parse("C:\\Users\\Admin\\AppData\\Local\\file.txt", None);
        assert_eq!(info.prefix, "C:\\");
        assert_eq!(info.filename, "file.txt");
        assert_eq!(info.segments[0].text, "Users");
        assert_eq!(info.segments[0].priority, SegmentPriority::Context);
        assert_eq!(info.segments[1].text, "Admin");
        assert_eq!(info.segments[1].priority, SegmentPriority::Identity);
    }

    #[test]
    fn parse_unc() {
        let info = PathInfo::parse("\\\\server\\share\\dept\\file.xlsx", None);
        assert_eq!(info.prefix, "\\\\server\\share");
        assert_eq!(info.filename, "file.xlsx");
        assert_eq!(info.segments.len(), 1);
        assert_eq!(info.segments[0].text, "dept");
    }

    #[test]
    fn parse_tilde() {
        let info = PathInfo::parse("~/projects/rust/file.rs", None);
        assert_eq!(info.prefix, "~");
        assert_eq!(info.filename, "file.rs");
        assert_eq!(info.segments.len(), 2);
    }

    #[test]
    fn parse_dot_backslash() {
        let info = PathInfo::parse(".\\Users\\Admin\\file.txt", None);
        assert_eq!(info.prefix, ".");
        assert_eq!(info.style, PathStyle::Windows);
        assert_eq!(info.filename, "file.txt");
    }

    #[test]
    fn parse_empty() {
        let info = PathInfo::parse("", None);
        assert_eq!(info.prefix, "");
        assert_eq!(info.filename, "");
        assert!(info.segments.is_empty());
    }

    #[test]
    fn parse_filename_only() {
        let info = PathInfo::parse("file.txt", None);
        assert_eq!(info.filename, "file.txt");
        assert!(info.segments.is_empty());
    }

    #[test]
    fn reassemble() {
        let info = PathInfo::parse("/home/john/projects/file.rs", None);
        let texts: Vec<&str> = info.segments.iter().map(|s| s.text.as_str()).collect();
        let result = info.reassemble(&texts);
        assert_eq!(result, "/home/john/projects/file.rs");
    }

    #[test]
    fn reassemble_windows() {
        let info = PathInfo::parse("C:\\Users\\Admin\\Docs\\file.txt", None);
        let texts: Vec<&str> = info.segments.iter().map(|s| s.text.as_str()).collect();
        let result = info.reassemble(&texts);
        assert_eq!(result, "C:\\Users\\Admin\\Docs\\file.txt");
    }
}

File: ts/src/path-info.ts
import { detectStyle, separator, UNIX_HOME_ROOTS, WIN_HOME_ROOTS } from './platform';
import type { PathStyle } from './platform';

export type SegmentPriority = 'sacred' | 'identity' | 'context' | 'expendable';

export interface Segment {
  text: string;
  priority: SegmentPriority;
}

export class PathInfo {
  constructor(
    public readonly prefix: string,
    public readonly segments: Segment[],
    public readonly filename: string,
    public readonly style: PathStyle,
  ) {}

  static parse(path: string, forceStyle?: PathStyle): PathInfo {
    const style = forceStyle ?? detectStyle(path);
    const sep = separator(style);

    const normalized =
      style === 'unix' ? path.replaceAll('\\', '/') : path.replaceAll('/', '\\');

    const [prefix, remainder] = extractPrefix(normalized, style);

    const parts = remainder.split(sep).filter(s => s !== '');

    if (parts.length === 0) {
      return new PathInfo(prefix, [], '', style);
    }

    const filename = parts[parts.length - 1];
    const dirParts = parts.slice(0, -1);

    const segments = classifySegments(dirParts, prefix, style);

    return new PathInfo(prefix, segments, filename, style);
  }

  reassemble(segmentTexts: string[]): string {
    const sep = separator(this.style);
    let result = this.prefix;

    for (let i = 0; i < segmentTexts.length; i++) {
      if (i > 0 || (result !== '' && !result.endsWith(sep))) {
        result += sep;
      }
      result += segmentTexts[i];
    }

    if (this.filename !== '') {
      if (result !== '' && !result.endsWith(sep)) {
        result += sep;
      }
      result += this.filename;
    }

    return result;
  }
}

function extractPrefix(path: string, style: PathStyle): [string, string] {
  if (style === 'windows') {
    if (path.startsWith('\\\\')) {
      const afterSlashes = path.slice(2);
      const serverEnd = afterSlashes.indexOf('\\');
      if (serverEnd !== -1) {
        const afterServer = afterSlashes.slice(serverEnd + 1);
        const shareEnd = afterServer.indexOf('\\');
        const actualShareEnd = shareEnd === -1 ? afterServer.length : shareEnd;
        const prefixEnd = 2 + serverEnd + 1 + actualShareEnd;
        const prefix = path.slice(0, prefixEnd);
        const remainder = prefixEnd < path.length ? path.slice(prefixEnd + 1) : '';
        return [prefix, remainder];
      }
      return [path, ''];
    }

    if (
      path.length >= 3 &&
      /^[a-zA-Z]$/.test(path[0]) &&
      path[1] === ':' &&
      path[2] === '\\'
    ) {
      return [path.slice(0, 3), path.slice(3)];
    }

    if (path.startsWith('.\\')) {
      return ['.', path.slice(2)];
    }

    if (path.startsWith('\\')) {
      return ['\\', path.slice(1)];
    }

    return ['', path];
  }

  if (path.startsWith('/')) {
    return ['/', path.slice(1)];
  }

  if (path.startsWith('~/')) {
    return ['~', path.slice(2)];
  }

  if (path === '~') {
    return ['~', ''];
  }

  return ['', path];
}

function classifySegments(
  parts: string[],
  prefix: string,
  style: PathStyle,
): Segment[] {
  const homeRoots = style === 'unix' ? UNIX_HOME_ROOTS : WIN_HOME_ROOTS;

  let identityIdx: number | null = null;
  if (
    parts.length > 0 &&
    homeRoots.some(root => parts[0].toLowerCase() === root.toLowerCase())
  ) {
    if (parts.length > 1) {
      identityIdx = 1;
    }
  }

  return parts.map((text, i) => {
    let priority: SegmentPriority;

    if (i === identityIdx) {
      priority = 'identity';
    } else if (
      i === 0 &&
      homeRoots.some(root => text.toLowerCase() === root.toLowerCase())
    ) {
      priority = 'context';
    } else {
      priority = 'expendable';
    }

    return { text, priority };
  });
}
